### PR TITLE
Cleanup some code in SortedSet

### PR DESCRIPTION
**Pull request by jamesqo**
- Rename and make `AreComparersEqual` an instance method
  - Check for reference equality before calling into `Equals()`
- Refactor large constructor body into smaller methods
- Add XML docs
- Add braces
- Remove `SortedSet<T>.` name qualification when it isn't necessary

/cc @ianhays @safern

### CODE REVIEW

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 1]
```diff
@@ -92,76 +92,40 @@ public SortedSet(IEnumerable<T> collection, IComparer<T> comparer)
...
-                while (theirCurrent != null)
-                {
-                    theirStack.Push(theirCurrent);
-                    myStack.Push(myCurrent);
-                    myCurrent.Left = (theirCurrent.Left != null ? new SortedSet<T>.Node(theirCurrent.Left.Item, theirCurrent.Left.IsRed) : null);
-                    theirCurrent = theirCurrent.Left;
-                    myCurrent = myCurrent.Left;
-                }
-                while (theirStack.Count != 0)
+            // These are explicit type checks in the mould of HashSet. It would have worked better with
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Nit: mould => mold

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 2]
```diff
@@ -92,76 +92,40 @@ public SortedSet(IEnumerable<T> collection, IComparer<T> comparer)
...
-            else
+            
+            int count;
+            T[] elements = EnumerableHelpers.ToArray(collection, out count);
+            if (count > 0)
             {
-                int count;
-                T[] els = EnumerableHelpers.ToArray(collection, out count);
-                if (count > 0)
+                Array.Sort(elements, 0, count, comparer);
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Why did you remove setting comparer to _comparer?  If comparer is null, aren't the below calls to comparer.Compare going to blow up?

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 3]
```diff
@@ -203,32 +171,40 @@ private bool ContainsAllElements(IEnumerable<T> collection)
...
             Node current = _root;
+
             while (current != null)
             {
                 stack.Push(current);
                 current = current.Left;
             }
-            while (stack.Count != 0)
+
+            while (stack.Count > 0)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Why change from != to >?

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 4]
```diff
@@ -968,10 +952,11 @@ internal static bool SortedSetEquals(SortedSet<T> set1, SortedSet<T> set2, IComp
 
         }
 
-        // This is a little frustrating because we can't support more sorted structures.
-        private static bool AreComparersEqual(SortedSet<T> set1, SortedSet<T> set2)
+        private bool HasEqualComparer(SortedSet<T> other)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Why did you rename this and make it instance instead of static?

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 5]
```diff
@@ -1796,19 +1781,67 @@ public Node(T item, bool isRed = true)
...
+                        newRight.Left = originalRight.Left?.ShallowClone();
+                        originalRight = originalRight.Left;
+                        newRight = newRight.Left;
+                    }
+                }
+
+                return newRoot;
+            }
+
+            public int GetCount() => 1 + (Left?.GetCount() ?? 0) + (Right?.GetCount() ?? 0);
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: debug only?

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 6]
```diff
@@ -1796,19 +1781,67 @@ public Node(T item, bool isRed = true)
...
+                while (originalCurrent != null)
+                {
+                    originalNodes.Push(originalCurrent);
+                    newNodes.Push(newCurrent);
+                    newCurrent.Left = originalCurrent.Left?.ShallowClone();
+                    originalCurrent = originalCurrent.Left;
+                    newCurrent = newCurrent.Left;
+                }
+
+                while (originalNodes.Count > 0)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Same question as elsewhere: why != to >?

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs (hunk 2 above)
Comment (jamesqo) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: You're correct, I was confused and thought comparer was only passed to `Array.Sort` which accepts null. I'll change it back.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs (hunk 3 above)
Comment (jamesqo) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: I prefer using > since `stack.Count` will always be positive. I can change it back though.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs (hunk 4 above)
Comment (jamesqo) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: In nearly all of the places this method is being used, `this` is passed as a parameter. This makes it more concise. I also think it's better to have instance methods when you can because then you don't have to worry about null in the callee.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs (hunk 5 above)
Comment (jamesqo) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: @stephentoub I can't do a `[Conditional("DEBUG")]` here because the return type is non-void, but I'll use an `#if` and throw if this method is called in Release.

**Comment by jamesqo** [comment]
Pushed a couple more changes (which can be viewed in the latest commit):

- Move some static methods handling nodes on `SortedSet` to instance methods on `Node`
- Capitalize comments, put backticks around parts that refer to code
- Prefer `IsBlack` to `!IsRed`
- Add curly braces around more statements
- Make optional parameter in Node constructor required
- Rename local variables in a few places

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 7]
```diff
@@ -110,6 +110,9 @@ public SortedSet(IEnumerable<T> collection, IComparer<T> comparer)
             T[] elements = EnumerableHelpers.ToArray(collection, out count);
             if (count > 0)
             {
+                // If `comparer` is null, sets it to Comparer<T>.Default. We checked for this condition in the IComparer<T> constructor.
+                // Array.Sort handles null comparers, but we need this later when we use `comparer.Compare` directly.
+                comparer = _comparer;
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Does the fact that tests passed previously without this suggest missing tests?

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 8]
```diff
@@ -1826,7 +1829,14 @@ public Node DeepClone(int count)
                 return newRoot;
             }
 
-            public int GetCount() => 1 + (Left?.GetCount() ?? 0) + (Right?.GetCount() ?? 0);
+            public int GetCount()
+            {
+#if !DEBUG
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: We shouldn't even build it into release builds.  We should just wrap the whole function and any usage in #if DEBUG.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 9]
```diff
@@ -335,8 +335,8 @@ internal virtual bool AddIfNotPresent(T item)
             Node grandParent = null;
             Node greatGrandParent = null;
 
-            // even if we don't actually add to the set, we may be altering its structure (by doing rotations
-            // and such). so update version to disable any enumerators/subsets working on it
+            // Even if we don't actually add to the set, we may be altering its structure (by doing rotations and such).
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: In the future, it'd be helpful to avoid doing lots of additional subsequent changes like this after the PR has already been reviewed.  It makes reviewing more difficult and time consuming.  Thanks.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 10]
```diff
@@ -575,20 +578,16 @@ public void CopyTo(T[] array, int index, int count)
...
             }
 
-            // upper bound
-            count += index;
+            count += index; // Make `count` the upper bound.
 
             InOrderTreeWalk(node =>
             {
-                if (index >= count)
+                if (index == count)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Why this change?

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 11]
```diff
@@ -665,53 +664,42 @@ void ICollection.CopyTo(Array array, int index)
...
             Debug.Assert(grandParent != null);
-            bool parentIsOnRight = (grandParent.Right == parent);
-            bool currentIsOnRight = (parent.Right == current);
+
+            bool parentIsRightChild = (grandParent.Right == parent);
+            bool currentIsRightChild = (parent.Right == current);
 
             Node newChildOfGreatGrandParent;
-            if (parentIsOnRight == currentIsOnRight)
+            if (parentIsRightChild == currentIsRightChild)
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: I personally find this name change makes it more confusing, not less.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 12]
```diff
@@ -1839,13 +1789,68 @@ public int GetCount()
...
+                return grandChild;
+            }
         }
 
         [SuppressMessage("Microsoft.Performance", "CA1815:OverrideEqualsAndOperatorEqualsOnValueTypes", Justification = "not an expected scenario")]
         [Serializable]
         public struct Enumerator : IEnumerator<T>, IEnumerator, ISerializable, IDeserializationCallback
         {
-            private static Node s_dummyNode = new Node(default(T));
+            private static Node s_dummyNode = new Node(default(T), isRed: true);
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: readonly?

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: > I also think it's better to have instance methods when you can because then you don't have to worry about null in the callee.

You still have to worry about null; you just push the responsibility to the compiler, e.g. generating a callvirt instead of a call.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs (hunk 10 above)
Comment (jamesqo) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: I can revert

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs (hunk 11 above)
Comment (jamesqo) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Will revert

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs (hunk 12 above)
Comment (jamesqo) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Yep

**Comment by danmosemsft** [comment]
@stephentoub any more feedback or good to go?
